subsystems/http/Http/Middleware/CsrfMiddleware.php

```diff
@@ -10,12 +12,54 @@
  */
 class CsrfMiddleware implements RequestHandlerInterface
 {
+
+  /**
+   * @var HttpSettings
+   */
+  private $httpSettings;
+  private $session;
+
+  /**
+   * CsrfMiddleware constructor.
+   *
+   * @param SessionInterface $session
+   * @param HttpSettings     $httpSettings
+   */
+  public function __construct (SessionInterface $session, HttpSettings $httpSettings)
+  {
+    $this->session = $session;
+    $this->httpSettings = $httpSettings;
+  }
+
   function __invoke (ServerRequestInterface $request, ResponseInterface $response, callable $next)
   {
-    /** @var ResponseInterface $response */
-    $response = $next();
+    $session = $this->session;
+
+    if($this->httpSettings->useCsrfToken ()){
+      if ($request->getMethod () == 'POST') {
+        $post = $request->getParsedBody ();
```

Comment:
Check if content-type header is `application/x-www-form-urlencoded` or `multipart/form-data`.
If it is, check the token; if it isn't, don't check the token.
BUT, it an `X-CSRF-Token` http header exists, use its value.